// exceljs/dist/es5/xlsx/xform/comment/r-xform.js
/**
 * Copyright (c) 2016 Guyon Roche
 * LICENCE: MIT - please refer to LICENCE file included with this module
 * or https://github.com/guyonroche/exceljs/blob/master/LICENSE
 */

'use strict';

var utils = require('../../../utils/utils');
var BaseXform = require('../base-xform');
var RowPropertiesXform = require('./rowproperties-xform')
var TextXForm = require('./text-xform')

var RXform = module.exports = function () {
  this.map = {
      rPr: new RowPropertiesXform(),
      t: new TextXForm()
  }
};

utils.inherits(RXform, BaseXform, {

  get tag() {
    return 'r';
  },

  render: function render(xmlStream, model) {
    if (model) {
      const _this = this;
      model.forEach(function(r) {
        xmlStream.openNode(_this.tag);
        _this.map.rPr.render(xmlStream, r.rPr);
        _this.map.t.render(xmlStream, r.t)
        xmlStream.closeNode();
      })
    }
  },

  parseOpen: function parseOpen(node) {
    if (this.parser) {
      this.parser.parseOpen(node);
      return true;
    }
    switch (node.name) {
      case this.tag:
        return true;
      default:
        this.parser = this.map[node.name];
        if (this.parser) {
          this.parser.parseOpen(node);
        }
        return true;
    }
  },
  parseText: function parseText(text) {
    if (this.parser) {
      this.parser.parseText(text);
      return true;
    }
  },
  parseClose: function parseClose(name) {
    if (name==this.tag) {
      var thisR = {}
      if (this.map.rPr.model) {
        thisR.rPr = this.map.rPr.model;
      }
      if (this.map.t.model) {
        thisR.t = this.map.t.model;
      }
      if (!this.model) this.model = []
      this.model.push(thisR);
      return false;
    }
    if (this.parser) {
      if (!this.parser.parseClose(name)) {
          this.parser = undefined;
      }
      return true;
    }
    return false;
  }
});
