// exceljs/dist/es5/xlsx/xform/comment/rowproperties-xform.js
/**
 * Copyright (c) 2016 Guyon Roche
 * LICENCE: MIT - please refer to LICENCE file included with this module
 * or https://github.com/guyonroche/exceljs/blob/master/LICENSE
 */

'use strict';

var utils = require('../../../utils/utils');
var BaseXform = require('../base-xform');

var RowPropertiesXform = module.exports = function () {};

utils.inherits(RowPropertiesXform, BaseXform, {

  get tag() {
    return 'rPr';
  },

  render: function render(xmlStream, model) {
    if (model) {
      xmlStream.openNode(this.tag);
      for (var obj in model) {
        if (model.hasOwnProperty(obj)) {
          xmlStream.leafNode(obj, model[obj]);
        } 
      }
      xmlStream.closeNode();
    }
  },

  parseOpen: function parseOpen(node) {
    if (this.parser) {
      this.parser.parseOpen(node);
      return true;
    }
    switch (node.name) {
      case this.tag:
        this.model = {}
        return true;
      default:
        this.model[node.name] = node.attributes
        return true;
    }
  },
  
  parseText: function parseText(text) {
    if (this.parser) {
      this.parser.parseText(text);
    }
  },
  parseClose: function parseClose(name) {
    if (name==this.tag){
      return false;
    } else {
      return true;
    }
  },
});
